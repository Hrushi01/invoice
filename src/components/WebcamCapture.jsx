import React from "react";
import Webcam from "react-webcam";
import IconButton from "@mui/material/IconButton";

import { Box, Button, Modal, Typography } from "@mui/material";

const videoConstraints = {
  width: 1280,
  height: 720,
  facingMode: "user",
};

const WebcamCapture = (props) => {
  const { camon, setCamOn, setpicture, picture } = props;
  const webcamRef = React.useRef(null);

  return (
    <>
      <div className="items-center">
        <Modal
          data-testid="webcam-modal"
          open={camon}
          onClose={() => {
            setCamOn(false);
          }}
        >
          <Box className="  left-1/2 right-1/2  translate-x-1/2 translate-y-1/3 bg-orange-50 opacity-90 border-2 border-gray-600 w-1/2 p-5  ">
            <div>
              <div className="  left-2 top-2">
                <IconButton
                  color="error"
                  data-testid="x-button"
                  component="label"
                  onClick={() => {
                    setCamOn(false);
                  }}
                >
                  X
                </IconButton>
              </div>
              <div
                className="flex justify-center text-center"
                data-testid="texte"
              >
                <Typography variant="h5">
                  Please have a clear Background.
                </Typography>
              </div>
              <div className="  p-1 flex justify-center text-center">
                <div className="p-1 border-2 border-gray-600 ">
                  {picture ? (
                    <div className="">
                      <img
                        data-testid="img"
                        src={
                          picture
                            ? picture
                            : "https://via.placeholder.com/150.png?text=Logo"
                        }
                        alt="img"
                        width={450}
                      />
                    </div>
                  ) : (
                    <Webcam
                      data-testid="webcam"
                      audio={false}
                      ref={webcamRef}
                      screenshotFormat="image/jpeg"
                      width={450}
                      // onUserMedia={onuser}
                      videoConstraints={videoConstraints}
                    />
                  )}
                </div>
              </div>
              <div className="grid grid-cols-3 gap-5  mt-4  ">
                <Button
                  data-testid="capture-button"
                  variant="contained"
                  onClick={() => {
                    const imageSrc = webcamRef.current.getScreenshot();
                    setpicture(imageSrc);
                  }}
                >
                  Capture
                </Button>
                <Button
                  className="retake"
                  data-testid="retake-button"
                  variant="contained"
                  onClick={() => setpicture(null)}
                >
                  Retake
                </Button>
                <Button
                  data-testid="close-button"
                  variant="contained"
                  color="error"
                  onClick={() => {
                    setCamOn(false);
                  }}
                >
                  Close
                </Button>
              </div>
            </div>
          </Box>
        </Modal>
      </div>
    </>
  );
};
export default WebcamCapture;
