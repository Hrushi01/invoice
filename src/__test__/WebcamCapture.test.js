import React from "react";
import { cleanup, screen } from "@testing-library/react";
import { render, unmountComponentAtNode } from "react-dom";

import { act } from "react-dom/test-utils";
import WebcamCapture from "../components/WebcamCapture";

afterEach(cleanup);

describe("WebcamCapture component", () => {
  let container;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it("renders webcam modal", () => {
    render(
      <WebcamCapture
        camon={true}
        setCamOn={() => {}}
        picture={null}
        setPicture={() => {}}
      />,
      container
    );
    const modal = screen.getByTestId("webcam-modal");
    expect(modal).toBeInTheDocument();
  });
  it("modal", () => {
    render(
      <WebcamCapture
        camon={true}
        setCamOn={() => {}}
        picture={null}
        setPicture={() => {}}
      />,
      container
    );
    expect(screen.getByTestId("webcam-modal")).toBeTruthy();
  });

  it("renders webcam", () => {
    render(
      <WebcamCapture
        camon={true}
        setCamOn={() => {}}
        picture={null}
        setPicture={() => {}}
      />,
      container
    );
    const webcam = screen.getByTestId("webcam");
    expect(webcam).toBeDefined();
  });

  it("captures webcam image", () => {
    const mockSetPicture = jest.fn();
    render(
      <WebcamCapture
        camon={true}
        setCamOn={() => {}}
        picture={null}
        setpicture={mockSetPicture}
      />,
      container
    );
    const captureButton = screen.getByTestId("capture-button");
    captureButton.click();
    expect(mockSetPicture).toHaveBeenCalled();
  });

  it("renders webc", () => {
    render(<WebcamCapture camon={true} />, container);
    expect(screen.getByTestId("webcam")).toBeTruthy();
  });

  it("renders image after", () => {
    render(<WebcamCapture camon={true} picture="test_image.jpeg" />, container);
    expect(screen.getByTestId("img")).toBeTruthy();
  });

  it("closes modal on clicking close button", () => {
    const setCamOn = jest.fn();
    render(<WebcamCapture camon={true} setCamOn={setCamOn} />, container);
    const closeBtn = screen.getByTestId("close-button");
    act(() => {
      closeBtn.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    expect(setCamOn).toHaveBeenCalledWith(false);
  });

  it("should call setPicture when clicked", () => {
    const setPicture = jest.fn();
    render(<WebcamCapture camon={true} setpicture={setPicture} />, container);

    const retakeButton = screen.getByTestId("retake-button");

    act(() => {
      retakeButton.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    expect(setPicture).toHaveBeenCalled();
  });
});
