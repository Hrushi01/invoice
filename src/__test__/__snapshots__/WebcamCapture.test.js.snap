// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebcamCapture component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="items-center"
  />
</DocumentFragment>
`;
